### DISCUSSION TITLE
Is there no problem when I use more than 16 characters for encryption?

### DISCUSSION THREAD

**Discussion by seareale** [pyinstaller/pyinstaller#7448] [PyInstaller]
I used pyinstaller to package my python program.
And as below, it was encrypted using more than 16 characters.

```
# build.spec

block_cipher = pyi_crypto.PyiBlockCipher(key='1234567890123456abcdefghijklmnop')
```

I tried to decompile exe file and found only 16 characters in the `pyimod00_crypto_key` file.
And even if I tried to decompile with the 16 characters, the decryption failed.
Also failed when using original key (`1234567890123456abcdefghijklmnop`).

#### in 'pyimod00_crypto_key'
<img width="118" alt="image" src="https://user-images.githubusercontent.com/33483699/220028792-a57fe8bc-e4ad-4a2c-a082-0d2ebd6e7029.png">

```
# using the key from 'pyimod00_crypto_key'

python pyinstxtractor.py PROGRAM.exe --key=1234567890123456
>[!] Error: Failed to decompress ....

python pyinstxtractor.py PROGRAM.exe --key=1234567890123456abcdefghijklmnop
>[!] Error: Failed to decompress ....
```

For this reason, can I use `--key` option for program security?

**Comment by rokm** [comment]
If you use more than 16 characters in your key, only first 16 are used ([here](https://github.com/pyinstaller/pyinstaller/blob/55eb4764f6c09894cd29180d5128476b8212fd21/PyInstaller/loader/pyimod01_archive.py#L211-L219) is the run-time code).

> ```
> python pyinstxtractor.py PROGRAM.exe --key=1234567890123456
> >[!] Error: Failed to decompress ....
> 
> python pyinstxtractor.py PROGRAM.exe --key=1234567890123456abcdefghijklmnop
> >[!] Error: Failed to decompress ....
> ```

Does `pyinstxtractor` even support the `--key` command-line option? From what I can tell, it is quietly ignoring it (which means it does not support bytecode decryption)...

> For this reason, can I use `--key` option for program security?

No. The key can be read from the executable by anyone familiar with PyInstaller's internal structures (i.e., anyone who can read PyInstaller's source) and used to decrypt the encrypted bytecode. And even though `pyinstxtractor` seemingly cannot do it, there are other tools that can (and it would not take much effort to add support to `pyinstxtractor`, either).

So this is mostly just smoke and mirrors, and therefore we are planning to remove this functionality altogether (see #6999).
